internal/datanode/importv2/executor.go

```diff
@@ -0,0 +1,330 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importv2
+
+import (
+	"fmt"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/samber/lo"
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/datanode/syncmgr"
+	"github.com/milvus-io/milvus/internal/proto/datapb"
+	"github.com/milvus-io/milvus/internal/proto/internalpb"
+	"github.com/milvus-io/milvus/internal/storage"
+	"github.com/milvus-io/milvus/internal/util/importutilv2"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/conc"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+)
+
+type Executor interface {
+	Start()
+	Slots() int64
+	Close()
+}
+
+type executor struct {
+	manager TaskManager
+	syncMgr syncmgr.SyncManager
+	cm      storage.ChunkManager
+
+	readConcurrently bool
+	pool             *conc.Pool[any]
+
+	closeOnce sync.Once
+	closeChan chan struct{}
+}
+
+func NewExecutor(manager TaskManager, syncMgr syncmgr.SyncManager, cm storage.ChunkManager) Executor {
+	pool := conc.NewPool[any](
+		paramtable.Get().DataNodeCfg.MaxConcurrentImportTaskNum.GetAsInt(),
+		conc.WithPreAlloc(false),
+		conc.WithDisablePurge(false),
+		conc.WithPreHandler(runtime.LockOSThread), // lock os thread for cgo thread disposal
+	)
+	return &executor{
+		manager:   manager,
+		syncMgr:   syncMgr,
+		cm:        cm,
+		pool:      pool,
+		closeChan: make(chan struct{}),
+	}
+}
+
+func (e *executor) Start() {
+	log.Info("start import executor")
+	var (
+		exeTicker = time.NewTicker(1 * time.Second)
+		logTicker = time.NewTicker(10 * time.Minute)
+	)
+	defer exeTicker.Stop()
+	defer logTicker.Stop()
+	for {
+		select {
+		case <-e.closeChan:
+			log.Info("import executor exited")
+			return
+		case <-exeTicker.C:
+			tasks := e.manager.GetBy(WithStates(internalpb.ImportState_Pending))
+			e.readConcurrently = true
+			if len(tasks) >= e.pool.Free()/2 {
+				e.readConcurrently = false
+			}
+			futures := make([]*conc.Future[any], 0, len(tasks))
+			for _, task := range tasks {
+				task := task
+				f := e.pool.Submit(func() (any, error) {
```

Comment:
Do not throw tasks into the pool here. The pool only caters to concurrent file reading operations within tasks.